### PR TITLE
Merge `SolutionTransfer` and `p::d::SolutionTransfer`

### PR DESCRIPTION
**Pull request by pcafrica**
Solves #15703.

It is also a necessary step before testing the `ParticleHandler` serialization with serial triangulations (@blaisb).

`T::{save,load}` can now handle cell attached data.

When calling `T::save`, four files are created:
- `filename.info` (containing version, n_proc, number of datasets attached, number of active cells)
- `filename_triangulation.data` (triangulation information - both coarse mesh and refinement information)
- `filename_fixed.data` (fixed-size cell attached data)
- `filename_variable.data` (variable-size cell attached data)

Open questions:
- [x] Should we move `p::d::SolutionTransfer` out of the `distributed` namespace?
- [x] Is the current use of `update_cell_relations` correct? There are many inconsistencies between `T`, `p::T` and `p::f::T`. I would say that we should call `update_cell_relations` by the end of each `create_triangulation` overload, but this would break some p4est tests.

FYI: @marcfehling @peterrum

### CODE REVIEW

File: source/grid/tria.cc
Comment (peterrum) [inline source/grid/tria.cc:16059]: This does not look right. I guess for load/save it does not matter, however, once actually interpolation is performed than we need to set this correctly.

**Review by peterrum — COMMENTED** [review]
Nice! Let's get the other PR merged soon!

**Comment by peterrum** [comment]
> Should we move p::d::SolutionTransfer out of the distributed namespace?

Not yet. I guess once we have the correct cell relations, we might be able to do the transfer between differently refined meshes also for the serial mesh so that we can replace `SolutionTransfer` by `p::d::SolutionTransfer`.

> but this would break some p4est tests.

What do you mean?

File: source/grid/tria.cc
Comment (pcafrica) [reply source/grid/tria.cc:13731]: Maybe a static variable in the new `CellAttachedDataSerializer` class?

File: source/grid/tria.cc
Comment (pcafrica) [reply source/grid/tria.cc:16059]: I agree.

Note: this implementation was copied by `p::f::T`. This means that `SolutionTransfer` does not currently support interpolation between `p::f::T`s.

**Comment by pcafrica** [comment]
@peterrum Consider the static `create_triangulation` method ([here](https://github.com/dealii/dealii/blob/58ee52a495a44cc354e947bfea82eb0f36676701/source/grid/tria.cc#L3237)).

I tried adding `tria.update_cell_relations()` as a very last line there, but this broke all the p4est parallel tests. They crashed with a segfault at [this line](https://github.com/dealii/dealii/blob/58ee52a495a44cc354e947bfea82eb0f36676701/source/distributed/tria.cc#L4008). My guess is that the `parallel_forest` member has not been fully initialized yet, so dereferencing it is invalid.

But this falls beyond my current understanding of things, so I will be investigating the issue further :)

**Comment by peterrum** [comment]
@pcafrica What is the status here?

**Comment by pcafrica** [comment]
No update so far. Unfortunately, I haven't been able to follow up on this PR due to other commitments.

**Comment by peterrum** [comment]
@pcafrica I have pushed here (https://github.com/peterrum/dealii/tree/serialization_solution_transfer) two commits how one could make `parallel::distributed::SolutionTransfer` for data transfer due to refinement/coarsening. See the test: https://github.com/peterrum/dealii/commit/19736def8f8192822ced81c22ebb94e50fa8ac2a#diff-d6583bc74d3e8b0a719b85ec8e32c2454c8f68bd2aaed1620403a1b96e8216e1.

File: tests/mpi/solution_transfer_07.cc
Comment (blaisb) [inline tests/mpi/solution_transfer_07.cc:68]: Any peculiar reason you changed the test here to print something new?

**Review by blaisb — APPROVED** [review]
Very very minor comments. God this unification will be great. Amazing work peter and thanks for taking the time to do this

File: doc/news/changes/incompatibilities/20240103AfricaBlaisMunch
Comment (peterrum) [inline doc/news/changes/incompatibilities/20240103AfricaBlaisMunch:10]: The change-log entry summarizes the PR.

**Comment by blaisb** [comment]
Ping, is there anyone  willing to give this another review?

File: source/grid/tria.cc
Comment (marcfehling) [inline source/grid/tria.cc:15829]: There is still a TODO here.

File: include/deal.II/grid/tria.h
Comment (marcfehling) [reply include/deal.II/grid/tria.h:3923]: I looked through the implementations of the three functions `update_cell_relations_serial`, `pack_data_serial_pre` and `pack_data_serial_post`. As far as I understood, all three of them manipulate the `cell_relations` object.

Can the three functions be unified into one? Ideally, an update of `cell_relations` should always give you the same result based on the currently set `refine` and `coarsen` flags, independent of the context when it is called.

File: include/deal.II/numerics/solution_transfer.h
Comment (marcfehling) [inline include/deal.II/numerics/solution_transfer.h:702]: Please add a `@deprecated` flag and point to the class that should be used instead.

Also, I noticed that you left the documentation for the `Legacy::SolutionTransfer` class, but removed it for `parallel::distributed::SolutionTransfer`. Is this on purpose?

File: source/numerics/solution_transfer.cc
Comment (marcfehling) [inline source/numerics/solution_transfer.cc:21]:
```suggestion
```
This should not be necessary.

File: include/deal.II/numerics/solution_transfer.h
Comment (marcfehling) [inline include/deal.II/numerics/solution_transfer.h:338]: This function does nothing. Do you want to keep it for compatibility purposes? If so, please deprecate it.

File: source/numerics/solution_transfer.cc
Comment (marcfehling) [inline source/numerics/solution_transfer.cc:507]: See above. There is no purpose to keep this function if it does nothing.

**Review by marcfehling — CHANGES_REQUESTED** [review]
That's a massive pull request. Thank you for working through it. I have a few comments.

I'm confused about the documentation of the SolutionTransfer class in its current state. Now it reads like it only works for parallel distributed applications. We need a more careful union of the documentation of both serial and distributed SolutionTransfer classes.

In the general description of the class, I suggest we split it in a part `General usage` and `Distributed grids`, where in the latter we describe that we need to pay attention whether vectors need to be ghosted etc.

**Comment by pcafrica** [comment]
Rebased on top of master.
I'm going to be working on the review soon. @peterrum @blaisb

File: source/numerics/solution_transfer.inst.in
Comment (peterrum) [reply source/numerics/solution_transfer.inst.in]: Apparently `dealii::LinearAlgebra::EpetraWrappers::Vector` is not working since there is no `operator[]`.

File: include/deal.II/grid/tria.h
Comment (peterrum) [reply include/deal.II/grid/tria.h]: > Can the three functions be unified into one?

I don't see how that would be possible. After refining/coarsening, the relevant information is lost.

File: include/deal.II/numerics/solution_transfer.h
Comment (peterrum) [reply include/deal.II/numerics/solution_transfer.h]: `parallel::distributed::SolutionTransfer` is now only an alias. The relevant information is now in the documentation of `SolutionTransfer`.

File: tests/hp/solution_transfer_16.cc
Comment (peterrum) [reply tests/hp/solution_transfer_16.cc:88]: `parallel::distributed::Triangulation` internally whether the values set by different cells are the same. If there are values in the vector, it needs assume that the value was previously set by another cell and an exception is raised.

**Comment by peterrum** [comment]
@marcfehling I have updated the PR as you suggested. Could you take another look! Thanks!

**Comment by bangerth** [comment]
I hope to work through this one last time in the next couple of weeks. Currently, there is a merge conflict here. Would you mind rebasing the patch?

I apologize for this taking so long! Let's get this merged soon!

File: examples/step-15/step-15.cc
Comment (bangerth) [inline examples/step-15/step-15.cc:409]: The commentary above this block is no longer accurate and needs to be adjusted.

Comment (bangerth) [reply examples/step-15/step-15.cc:409]: And if we don't need `current_solution` any more, we can probably get rid of the `tmp` vector as well.

**Review by bangerth — COMMENTED** [review]
Let's take the tutorials separately. If I understand this right, then the changes made to the tutorials are correct even without the rest of the patch, right?

If so, would one of you be willing to take these parts of the patch and move them to a separate PR? I think we can merge them independently relatively easily, subject to just the one comment about adjust a comment.

File: source/grid/tria.cc
Comment (bangerth) [reply source/grid/tria.cc:15836]: Of course, it could also be that the function is simply poorly named and I'm misunderstanding what it does. Perhaps a better name would be something like `do_coarsening_and_refinement_preactions()`, i.e., do some stuff that needs to happen before we refine/coarsen. And the other function would be `do_coarsening_and_refinement_postactions()`.

File: source/grid/tria.cc
Comment (bangerth) [inline source/grid/tria.cc:16015]: This seems to support my suspicion that the functions mentioned above are perhaps not named in a way that reveals what they do.

File: source/numerics/solution_transfer.inst.in
Comment (bangerth) [inline source/numerics/solution_transfer.inst.in:33]: You're listing specific vector types here. Why not leave it a loop over all vector types? Does the class not work for specific vector types? If so, which are these?

**Review by bangerth — COMMENTED** [review]
Here are a bunch of comments about `tria.h/cc`. I must admit that I have not tried to fully understand the logic and instead look for more formal criteria. I'll get to the `solution_transfer.h/cc` files in the next couple of days.

**Review by marcfehling — APPROVED** [review]
The logic of the functions `Triangulation::pack(), unpack(), update_cell_relations()` originally stems from the DataTransfer class, which was an internal class of parallel::distributed::Triangulation. Over time and continuous efforts to generalize the functionality, the interface migrated further and further to parent classes. I am okay with the current state and interface, but in the long run, it would probably make sense to introduce a DataTransferPolicy or SolutionTransferPolicy class (similar to DoFHandlerPolicy) that contains the logic that is now split between `pack()` and `pack_serial()` etc. Maybe as part of #15280.

The general documentation of the SolutionTransfer class is now closer to the documentation of the parallel::distributed::SolutionTransfer class. @bangerth, is there anything from the original SolutionTransfer documentation that you would like to keep?

Lots of changes in the tests make me worried about hard incompatibilities. But I think that's okay with the comments in the changelog.

Who is in charge of the PR now? I can help and contribute by pushing directly to this branch. Maybe we can speed things up a bit this way.

**Comment by peterrum** [comment]
> I can help and contribute by pushing directly to this branch. Maybe we can speed things up a bit this way.

@marcfehling  Feel free to push directly!

**Comment by pcafrica** [comment]
Hi all! I just tried to rebase and to fix explicit instantiations for all Vector types (#16880 #16902).

Though, there is still a lot to do to before having this PR mergeable :) How should we split up work?

@peterrum @marcfehling

**Comment by bangerth** [comment]
@pcafrica Are there pieces that could be broken out of this patch that could be merged separately? I would love for this to make it into the upcoming release. If we can find ways to make that process easier, I'm all for it!

**Comment by pcafrica** [comment]
@bangerth My two cents is that this PR should be merged as a whole. The main modifications are in the `*solution_transfer*` files, everything else is only adapting tutorials and tests to the new interface.

@peterrum What's your opinion?

File: source/grid/tria.cc
Comment (pcafrica) [reply source/grid/tria.cc]: Done!

**Comment by bangerth** [comment]
What's the status here? 

@peterrum Any comments on @pcafrica 's last question?

@pcafrica The `indent` check failed, but the log is no longer stored. Can you rebase and force push?

**Comment by peterrum** [comment]
Sorry for the later reply. I currently don't have time to implement the restructuring requested by @bangerth. I would suggest to continue here during the workshop and after the release (so that we don't merge such a large incompatibility a few days before release).

**Comment by peterrum** [comment]
I have rebased and all tests are still passing. Reminder to myself: 
```
grep -rl "SolutionTransfer" | while read -r line ; do printf ${line::-3}"|"; done
```

I have adjusted the milestone. @pcafrica @bangerth @blaisb @marcfehling We should get this in during the workshop (which is soon).

File: source/grid/tria.cc
Comment (pcafrica) [reply source/grid/tria.cc]: Actually the `file_basename` variable is used by appending suffixes for different files, so unless I'm missing something a single `std::ostream` would not suffice.

File: source/grid/tria.cc
Comment (pcafrica) [reply source/grid/tria.cc]: Done

File: source/grid/tria.cc
Comment (pcafrica) [reply source/grid/tria.cc]: Done

File: source/grid/tria.cc
Comment (pcafrica) [reply source/grid/tria.cc]: Done

File: source/grid/tria.cc
Comment (pcafrica) [reply source/grid/tria.cc]: See above comment.

File: source/grid/tria.cc
Comment (marcfehling) [inline source/grid/tria.cc:15842]: What is the purpose of `unpack_data_serial` here?

The function just updates the `local_cell_relations`, right? But you call `update_cell_relations()` already in line 15868.

**Review by marcfehling — APPROVED** [review]
Please rebase on the current master branch. We now run all examples as tests in the testsuite -- it would be nice to double check.

Otherwise looks okay to me after the remaining points have been addressed. I must admit though that I am starting to become kind of 'snow-blind' to reviewing SolutionTransfer code.

**Review by marcfehling — APPROVED** [review]
I'm fine with merging as is.

:shipit:

**Comment by marcfehling** [comment]
```
The following tests FAILED:

	3527 - parameter_handler/parameter_handler_26.debug (Failed)

Errors while running CTest
```

This test failure has clearly nothing to do with this PR.

**Comment by blaisb** [comment]
CELEBRATE GOOD TIMES COME ON!

**Pull request merged**
